components/calendar/src/cal/hijri.rs

```diff
@@ -66,109 +74,339 @@ fn era_year(year: i32) -> EraYear {
+                    29 => false,
+                    30 => true,
+                    31 => {
+                        icu_provider::log::trace!(
+                            "({}) Found year {monotonic_year} AH with month length {days_in_month} for month {}.",
+                            self.debug_name(),
+                            month_idx + 1
+                        );
+                        excess_days += 1;
+                        true
+                    }
+                    _ => {
+                        debug_assert!(
+                            !WELL_BEHAVED_ASTRONOMICAL_RANGE.contains(&start_day),
+                            "({}) Found year {monotonic_year} AH with month length {days_in_month} for month {}!",
+                            self.debug_name(),
+                            month_idx + 1
+                        );
+                        false
+                    }
+                }
+            });
+            // To maintain invariants for calendar arithmetic, if astronomy finds
+            // a 31-day month, "move" the day to the first 29-day month in the
+            // same year to maintain all months at 29 or 30 days.
+            if excess_days != 0 {
+                debug_assert!(
+                    excess_days == 1 || !WELL_BEHAVED_ASTRONOMICAL_RANGE.contains(&start_day),
+                    "({}) Found year {monotonic_year} AH with more than one excess day!",
+                    self.debug_name()
+                );
+                if let Some(l) = month_lengths.iter_mut().find(|l| !(**l)) {
+                    *l = true;
+                }
+            }
+            month_lengths
+        };
+        (start_day, month_lengths)
     }
 }
 
-/// The [Umm al-Qura Hijri Calendar](https://en.wikipedia.org/wiki/Islamic_calendar#Saudi_Arabia's_Umm_al-Qura_calendar)
-///
-/// This calendar is the official calendar in Saudi Arabia.
-///
-/// # Era codes
+/// The [Umm al-Qura Hijri sighting](https://en.wikipedia.org/wiki/Islamic_calendar#Saudi_Arabia's_Umm_al-Qura_calendar)
 ///
-/// This calendar uses two era codes: `ah`, and `bh`, corresponding to the Anno Hegirae and Before Hijrah eras
-///
-/// # Month codes
-///
-/// This calendar is a pure lunar calendar with no leap months. It uses month codes
-/// `"M01" - "M12"`.
+/// This is the official calendar in Saudi Arabia.
 #[derive(Clone, Debug, Default)]
 #[non_exhaustive]
-pub struct HijriUmmAlQura;
+pub struct UmmAlQura;
 
-/// The [tabular Hijri Calendar](https://en.wikipedia.org/wiki/Tabular_Islamic_calendar).
-///
-/// See [`HijriTabularEpoch`] and [`HijriTabularLeapYears`] for customization.
-///
-/// The most common version of this calendar uses [`HijriTabularEpoch::Friday`] and [`HijriTabularLeapYears::TypeII`].
-///
-/// # Era codes
-///
-/// This calendar uses two era codes: `ah`, and `bh`, corresponding to the Anno Hegirae and Before Hijrah eras
-///
-/// # Month codes
+impl HijriSighting for UmmAlQura {
```

Comment:
Question/Suggestion: Do we check the invariant that the number of days in the year equals the number of days between this new year RD and the RD of the new year that follows?